rotation.go

```diff
@@ -312,54 +312,64 @@ func (b *backend) setStaticAccountPassword(ctx context.Context, s logical.Storag
 	// Re-use WAL ID if present, otherwise PUT a new WAL
 	output := &setStaticAccountOutput{WALID: input.WALID}
 
+	b.Lock()
+	defer b.Unlock()
+
 	config, err := readConfig(ctx, s)
 	if err != nil {
-		return nil, err
+		return output, err
 	}
 	if config == nil {
-		return nil, errors.New("the config is currently unset")
-	}
-
-	newPassword, err := b.GeneratePassword(ctx, config)
-	if err != nil {
-		return nil, err
+		return output, errors.New("the config is currently unset")
 	}
 
-	oldPassword := input.Role.StaticAccount.Password
+	var newPassword string
+	if output.WALID != "" {
+		wal, err := b.findStaticWAL(ctx, s, output.WALID)
+		if err != nil {
+			return output, errwrap.Wrapf("error retrieving WAL entry: {{err}}", err)
+		}
 
-	// Take out the backend lock since we are swapping out the connection
-	b.Lock()
-	defer b.Unlock()
+		switch {
+		case wal != nil && wal.NewPassword != "":
+			newPassword = wal.NewPassword
+		default:
+			if wal == nil {
+				b.Logger().Error("expected role to have WAL, but WAL not found in storage", "role", input.RoleName, "WAL ID", output.WALID)
+			} else {
+				b.Logger().Error("expected WAL to have a new password set, but empty", "role", input.RoleName, "WAL ID", output.WALID)
+				err = framework.DeleteWAL(ctx, s, output.WALID)
+				if err != nil {
+					b.Logger().Warn("failed to delete WAL with no new password", "error", err, "WAL ID", output.WALID)
+				}
+			}
+			// If there's anything wrong with the WAL in storage, we'll need
+			// to generate a fresh WAL and password
+			output.WALID = ""
+		}
+	}
 
 	if output.WALID == "" {
+		newPassword, err = b.GeneratePassword(ctx, config)
+		if err != nil {
+			return output, err
+		}
 		output.WALID, err = framework.PutWAL(ctx, s, staticWALKey, &setCredentialsWAL{
 			RoleName:          input.RoleName,
 			Username:          input.Role.StaticAccount.Username,
 			DN:                input.Role.StaticAccount.DN,
 			NewPassword:       newPassword,
-			OldPassword:       oldPassword,
 			LastVaultRotation: input.Role.StaticAccount.LastVaultRotation,
 		})
 		if err != nil {
 			return output, errwrap.Wrapf("error writing WAL entry: {{err}}", err)
 		}
+		b.Logger().Debug("writing WAL", "role", input.RoleName, "WAL ID", output.WALID)
 	}
 
 	// Update the password remotely.
 	if err := b.client.UpdatePassword(config.LDAP, input.Role.StaticAccount.DN, newPassword); err != nil {
```

Comment:
Given all the branching above can we verify `len(newPassword) > 0` before updating the PW? It seems fine now, but this ought to future proof the code a bit

Reply:
Seems reasonable. In terms of handling that situation, I've regenerated the password and logged an error. That relies on the generate password function never returning an empty string, because we don't check the new password again, but I hope that's a reasonable assumption.